test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,172 @@
+class EchoServerImpl final : public grpc::testing::TestService::Service {
+  Status EmptyCall(::grpc::ServerContext* context,
+                   const grpc::testing::Empty* request,
+                   grpc::testing::Empty* response) {
+    srand(unsigned(time(0)));
+    unsigned int server_delay_microseconds = 100000;
+    server_delay_microseconds += rand() % server_delay_microseconds;
+    usleep(server_delay_microseconds);
+    return Status::OK;
+  }
+};
+
+// Run server in a thread
+void RunServer() {
+  // register channelz service
+  ::grpc::channelz::experimental::InitChannelzService();
+
+  EchoServerImpl service;
+  grpc::EnableDefaultHealthCheckService(true);
+  grpc_impl::reflection::InitProtoReflectionServerBuilderPlugin();
+  grpc::ServerBuilder builder;
+  auto server_creds =
+      grpc::testing::GetCredentialsProvider()->GetServerCredentials(
+          custom_credentials_type);
+  builder.AddListeningPort(server_address, server_creds);
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  gpr_log(GPR_INFO, "Server listening on %s", server_address.c_str());
+  server->Wait();
+}
+
+// Creata an echo client - set timeout as 0.15s
+class EchoClientImpl {
+ public:
+  EchoClientImpl(std::shared_ptr<Channel> channel)
+      : stub_(grpc::testing::TestService::NewStub(channel)) {}
+  Status EmptyCall() {
+    grpc::testing::Empty request;
+    grpc::testing::Empty response;
+    ClientContext context;
+    int64_t timeout_microseconds = 150;
+    context.set_deadline(
+        grpc_timeout_milliseconds_to_deadline(timeout_microseconds));
+    Status status = stub_->EmptyCall(&context, request, &response);
+    return status;
+  }
+
+ private:
+  std::unique_ptr<grpc::testing::TestService::Stub> stub_;
+};
+
+// Run client in a thread
+void RunClient(std::string client_id) {
+  // std::string target_str = "localhost:10000";
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  EchoClientImpl echoer(grpc::CreateChannel(server_address, channel_creds));
+  unsigned int client_echo_sleep_second = 1;
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
+    Status status = echoer.EmptyCall();
+    sleep(client_echo_sleep_second);
+  }
+}
+
+// Test the channelz sampler
+TEST(ChannelzSamplerTest, SimpleTest) {
+  // server thread
+  std::thread server_thread(RunServer);
+  gpr_log(GPR_INFO, "Wait 3 seconds to make sure server is started...");
+  float server_start_seconds = 3.0;
+  sleep(server_start_seconds);
```

Comment:
instead of sleeping for 3 seconds, can we create a channel that tries to connect to the server, and returns when it does connect, using the `WaitForConnected` channel API?

E.g., https://github.com/grpc/grpc/blob/master/test/cpp/end2end/grpclb_end2end_test.cc#L959

Reply:
Done. Thanks for pointing out.